### PR TITLE
[5.7] Fix:assertExactJson() does not work for empty JSON objects

### PR DESCRIPTION
**Pull request by alberthuang24**
related: https://github.com/laravel/framework/issues/25769

### CODE REVIEW

File: src/Illuminate/Support/helpers.php [hunk 1]
```diff
@@ -1160,3 +1160,28 @@ function with($value, callable $callback = null)
...
     }
 }
+
+if (! function_exists('merge_associate_array_recursive')) {
+    /**
+     * @param array $a
+     * @param array $b
+     * @return array
+     */
+    function merge_associate_array_recursive(array $a, array $b)
```
Comment (mfn) [inline src/Illuminate/Support/helpers.php]: No more global helper/functions please 👎

Comment (HongjiangHuang) [reply src/Illuminate/Support/helpers.php]: ok

File: src/Illuminate/Foundation/Testing/TestResponse.php [hunk 2]
```diff
@@ -685,17 +703,22 @@ public function assertJsonMissingValidationErrors($keys)
      */
     public function decodeResponseJson($key = null)
     {
-        $decodedResponse = json_decode($this->getContent(), true);
+        $decodedResponseStdClass = (array) json_decode($this->getContent());
+        $decodedResponseAssoc = json_decode($this->getContent(), true);
```
Comment (lucasmichot) [inline src/Illuminate/Foundation/Testing/TestResponse.php]: This should be moved just before `$decodedResponse =` as we might fail/throw an exception.

File: src/Illuminate/Foundation/Testing/TestResponse.php [hunk 3]
```diff
@@ -685,17 +703,22 @@ public function assertJsonMissingValidationErrors($keys)
...
                 throw $this->exception;
             } else {
                 PHPUnit::fail('Invalid JSON was returned from the route.');
             }
         }
 
-        return data_get($decodedResponse, $key);
+        $decodedResponse = static::mergeAssociateArrayRecursive($decodedResponseStdClass, $decodedResponseAssoc);
+
+        $data = data_get($decodedResponse, $key);
```
Comment (lucasmichot) [inline src/Illuminate/Foundation/Testing/TestResponse.php]: Why not just `return data_get($decodedResponse, $key);` ?

File: src/Illuminate/Foundation/Testing/TestResponse.php (hunk 2 above)
Comment (HongjiangHuang) [reply src/Illuminate/Foundation/Testing/TestResponse.php]:
```
$decodedResponseAssoc = json_decode($this->getContent(), true);
$decodedResponseStdClass = (array) json_decode($this->getContent());
```
such ?

Comment (lucasmichot) [reply src/Illuminate/Foundation/Testing/TestResponse.php]: `$decodedResponseAssoc` is useless in this context and should be moved down

File: src/Illuminate/Foundation/Testing/TestResponse.php [hunk 4]
```diff
@@ -57,6 +57,24 @@ public static function fromBaseResponse($response)
         return new static($response);
     }
 
+    public static function mergeAssociateArrayRecursive(array $a, array $b): array
```
Comment (phannaly) [inline src/Illuminate/Foundation/Testing/TestResponse.php]: Should have docblock

File: src/Illuminate/Foundation/Testing/TestResponse.php [hunk 5]
```diff
@@ -57,6 +57,30 @@ public static function fromBaseResponse($response)
...
+     * @param  array $arrayStdClass
+     * @param  array $arrayAssoc
+     * @return array
+     */
+    public static function mergeAssociateArrayRecursive(array $arrayStdClass, array $arrayAssoc): array
+    {
+        $merged = $arrayStdClass;
+
+        foreach ($arrayAssoc as $k => $v) {
+            if (! empty($v)) {
```
Comment (mfn) [inline src/Illuminate/Foundation/Testing/TestResponse.php]: Prefer early continue, less indentation for the following code:
```php
if (empty($v)) {
  continue;
}

if (…
```

**Comment by mfn** [comment]
I'm torn. Guess it fixes the issue which is valid in my book. The solution looks ... cryptic. I don't have anything better to offer though :|

But: please add inline documentation _why_ we've stdClass/Assoc stuff in there; a clear explaination should be left wth is going on here.

With this PR: yes, you understand it.
But: if you view the function in its completeness => no one will understand why. 😄

**Comment by HongjiangHuang** [comment]
@mfn Hi. ` json_decode ` if the second parameter to true will put stdClass is convert  to array.
If the assoc array is empty, PHP cannot tell if it was originally stdClass or array.
So, ` $decodedResponseStdClass = (array) json_decode ($this - > getContent ()); ` to obtain the original data ( StdClass will be retained here, ` $decodedResponseAssoc = json_decode ($this - > getContent (), true); `  to obtain the assoc array.


Such as: 
```
$emptyJsonObject = json_encode((object) []);   //"{}"

$a = json_decode($emptyJsonObject);   //object(stdClass)#1 (0) {}

$b = json_decode($emptyJsonObject, true);  //array(0) {}

$result = empty($b) ?  $a : $b;   // object(stdClass)#1 (0) {}
```

```
$emptyJsonObject = json_encode(['a' => 1]);   // "{"a":1}"

$a = json_decode($emptyJsonObject);   //object(stdClass)#1 (0) {"a" => 1}

$b = json_decode($emptyJsonObject, true);  //array(0) {"a" => 1}

$result = empty($b) ?  $a : $b;  //array(0) {"a" => 1}
```

If `$b` is not empty, overwrite a `$a`. The result is what we want.

**Comment by HongjiangHuang** [comment]
@mfn I haven't written the inline document you mentioned before. Do you have any examples for reference

**Comment by taylorotwell** [comment]
I am closing this pull request because it lacks sufficient explanation, tests, or both. It is difficult for us to merge pull requests without these things because the change may introduce breaking changes to the framework.

Feel free to re-submit your change with a **thorough** explanation of the feature and tests - integration tests are preferred over unit tests. Please include it's benefit to end users; the reasons it does not break any existing features; how it makes building web applications easier, etc.

Thanks!

**Pull request closed without merging**